### ISSUE TITLE
Golden Sun sprite priority

### ISSUE THREAD

**Issue by paulb-nl** [MiSTer-devel/GBA_MiSTer#38]
Reported here with the Japanese version: [http://www.atari-forum.com/viewtopic.php?p=389976#p389976](http://www.atari-forum.com/viewtopic.php?p=389976#p389976)

[Save file](https://github.com/MiSTer-devel/GBA_MiSTer/files/4014144/Ougon.no.Taiyou.-.Ushinawareshi.Toki.Japan.sav.zip)

![l3RYygl](https://user-images.githubusercontent.com/2227984/71642336-1664f900-2caa-11ea-90d1-f5d530b6323c.png)
![JQVNHRU](https://user-images.githubusercontent.com/2227984/71642338-18c75300-2caa-11ea-9bbe-09a599aec47f.png)

This is some interesting behavior. The ice is BG1/2 with priority 2 and the player sprite is sprite 0 with priority 3.

However when the player should be on top of the ice then the game places normal sprites with transparent tiles with priority 2 on top of sprite 0 at sprite 2/3. This changes the priority of sprite 0 to priority 2.

I have (poorly) hacked bld_demo.gba with a hex editor to do the same thing and tested it on GBA hardware and got the same behavior.

Note that this also changes the Semi-Transparent pixels to normal pixels because the blending on the sprite in bld_demo.gba does not happen anymore when Special effects are disabled. mGBA almost has the correct behavior except for this part.

So in summary when a sprite overlaps a sprite with lower priority then the GBA overwrites the priority and Semi-Transparent bit but not the pixels if they are transparent.

**Comment by RobertPeip** [comment]
That's really interesting behavior. To do that, we would need to seperate the color data from the priority/alpha data for the sprite drawing and save them seperatly until mixing.

**Comment by RobertPeip** [comment]
I did a quick shot with my emulator: if i allow updating prio and alpha even when the pixel is transparent, it works with the ice. I could implement that in the FPGA, too. However i'm not sure if it will break anything....

**Comment by RobertPeip** [comment]
Paul, are you going it fix it or should i take a look. Just in case we don't do double work.
Thanks again for the in depth analysis, great work!

**Comment by paulb-nl** [comment]
I am not working on it currently so you can go ahead!

**Comment by RobertPeip** [comment]
fixed

**Comment by paulb-nl** [comment]
Nice work Robert! I have done some testing and everything seems to be working fine.

The Tonc prio_demo.gba is also affected by this and is now identical to GBA hardware when sprite sorting is disabled by pressing select. Some red sprites are pulled in front of the background.

![20200107_221352-prio_demo2x](https://user-images.githubusercontent.com/2227984/71937486-c1931800-31ac-11ea-9488-d8b34701b2fe.png)

**Comment by endrift** [comment]
Can I get that hacked up bld demo ROM that's broken in mGBA, or a savestate of Golden Sun?

**Comment by paulb-nl** [comment]
Here is the bld_demo hack. I have added a screenshot with expected behavior: sprite behind the background, a small part on the left above the background and Semi-Transparency on mode 0 should not happen on that part. 

mGBA is only incorrect on the Semi-Transparency part.

[bld_demo(prio_hack).zip](https://github.com/MiSTer-devel/GBA_MiSTer/files/4042674/bld_demo.prio_hack.zip)
